components/brave_wallet/browser/zcash/zcash_resolve_transaction_status_task.cc

```diff
@@ -0,0 +1,168 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_resolve_transaction_status_task.h"
+
+#include <utility>
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_tx_meta.h"
+
+namespace brave_wallet {
+
+namespace {
+constexpr size_t kTransactionCompleteConfirmations = 3u;
+constexpr size_t kTransactionExpiryHours = 2u;
+
+using ResolveTransactionStatusResult =
+    ZCashWalletService::ResolveTransactionStatusResult;
+
+}  // namespace
+
+ZCashResolveTransactionStatusTask::ZCashResolveTransactionStatusTask(
+    absl::variant<base::PassKey<ZCashWalletService>,
+                  base::PassKey<class ZCashResolveTransactionStatusTaskTest>>
+        pass_key,
+    ZCashActionContext context,
+    ZCashWalletService& zcash_wallet_service,
+    std::unique_ptr<ZCashTxMeta> tx_meta,
+    ZCashResolveTransactionStatusTaskCallback callback)
+    : context_(std::move(context)),
+      zcash_wallet_service_(zcash_wallet_service),
+      tx_meta_(std::move(tx_meta)),
+      callback_(std::move(callback)) {}
+
+ZCashResolveTransactionStatusTask::~ZCashResolveTransactionStatusTask() =
+    default;
+
+void ZCashResolveTransactionStatusTask::Start() {
+  DCHECK(!started_);
+  started_ = true;
+  ScheduleWorkOnTask();
+}
+
+void ZCashResolveTransactionStatusTask::ScheduleWorkOnTask() {
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE, base::BindOnce(&ZCashResolveTransactionStatusTask::WorkOnTask,
+                                weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashResolveTransactionStatusTask::WorkOnTask() {
+  if (error_) {
+    std::move(callback_).Run(base::unexpected(error_.value()));
+    zcash_wallet_service_->ResolveTransactionStatusTaskDone(this);
+    return;
+  }
+
+  if (!tx_meta_ || !tx_meta_->tx()) {
+    std::move(callback_).Run(ResolveTransactionStatusResult::kExpired);
+    zcash_wallet_service_->ResolveTransactionStatusTaskDone(this);
+    return;
+  }
+
+  if (!chain_tip_) {
+    GetChainTip();
+    return;
+  }
+
+  if (!transaction_) {
+    GetTransaction();
+    return;
+  }
+
+  CHECK(tx_meta_);
+  CHECK(tx_meta_->tx());
+  CHECK(chain_tip_.value());
+  CHECK(transaction_.value());
+
+  uint32_t chain_tip = chain_tip_.value()->height;
+  uint32_t tx_height = transaction_.value()->height;
+  uint32_t expiry_height = tx_meta_->tx()->expiry_height();
+
+  // https://github.com/zcash/lightwalletd/blob/339b6d37e839d27bbd167ed02627ab7ab4d7051f/walletrpc/service.proto#L53
+  if (tx_height == 0 || tx_height == 0xFFFFFFFF) {
+    bool expired = false;
+    if (expiry_height != 0) {
+      expired = chain_tip > expiry_height;
+    } else {
+      // Fallback to the submitted time
+      expired = base::Time::Now() - tx_meta_->submitted_time() >
+                base::Hours(kTransactionExpiryHours);
+    }
```

Comment:
base::Time::Now() is not guaranteed to be increasing. Let's guard it so substraction is always correct